crates/next-api/src/nft_json.rs

```diff
@@ -0,0 +1,191 @@
+    /// Additional assets to include in the nft json. This can be used to manually collect assets
+    /// that are known to be required but are not in the graph yet, for whatever reason.
+    ///
+    /// An example of this is the two-phase approach used by the `ClientReferenceManifest` in
+    /// next.js.
+    additional_assets: Vec<ResolvedVc<Box<dyn OutputAsset>>>,
+}
+
+#[turbo_tasks::value_impl]
+impl NftJsonAsset {
+    #[turbo_tasks::function]
+    pub fn new(
+        chunk: Vc<Box<dyn OutputAsset>>,
+        output_fs: Vc<DiskFileSystem>,
+        project_fs: Vc<DiskFileSystem>,
+        client_fs: Vc<Box<dyn FileSystem>>,
+        additional_assets: Vec<ResolvedVc<Box<dyn OutputAsset>>>,
+    ) -> Vc<Self> {
+        NftJsonAsset {
+            chunk,
+            output_fs,
+            project_fs,
+            client_fs,
+            additional_assets,
+        }
+        .cell()
+    }
+}
+
+#[turbo_tasks::value(transparent)]
+pub struct OutputSpecifier(Option<RcStr>);
+
+#[turbo_tasks::value_impl]
+impl NftJsonAsset {
+    #[turbo_tasks::function]
+    async fn ident_in_project_fs(self: Vc<Self>) -> Result<Vc<FileSystemPath>> {
+        let this = self.await?;
+        let project_fs = this.project_fs.await?;
+        let output_fs = this.output_fs.await?;
+        let nft_folder = self.ident().path().parent().await?;
+
+        if let Some(subdir) = output_fs.root.strip_prefix(&*project_fs.root) {
+            Ok(this
+                .project_fs
+                .root()
+                .join(subdir.into())
+                .join(nft_folder.path.clone()))
+        } else {
+            // TODO: what are the implications of this?
+            bail!("output fs not inside project fs");
+        }
+    }
+
+    #[turbo_tasks::function]
+    async fn ident_in_client_fs(self: Vc<Self>) -> Result<Vc<FileSystemPath>> {
+        Ok(self
+            .await?
+            .client_fs
+            .root()
+            .join(self.ident().path().parent().await?.path.clone()))
+    }
+
+    #[turbo_tasks::function]
+    async fn get_output_specifier(
+        self: Vc<Self>,
+        path: Vc<FileSystemPath>,
+    ) -> Result<Vc<OutputSpecifier>> {
+        let this = self.await?;
+        let path_fs = path.fs().resolve().await?;
+        let path_ref = path.await?;
+        let nft_folder = self.ident().path().parent().await?;
+
+        if path_fs == Vc::upcast(this.output_fs.resolve().await?) {
+            // e.g. a referenced chunk
+            return Ok(Vc::cell(Some(
+                nft_folder.get_relative_path_to(&path_ref).unwrap(),
+            )));
+        } else if path_fs == Vc::upcast(this.project_fs.resolve().await?) {
+            return Ok(Vc::cell(Some(
+                self.ident_in_project_fs()
+                    .await?
+                    .get_relative_path_to(&path_ref)
+                    .unwrap(),
+            )));
+        } else if path_fs == Vc::upcast(this.client_fs.resolve().await?) {
```

Comment:
Why would the server reference assets on client side? Even if it references them, we don't need to include them in nft files, do we?

Reply:
If I recall correctly, some `next/font` font functionality causes this. And maybe edge is related here as well. I'd have to run this with logging aign.